File: ThalesMobileSimulator/ThalesCoreSource/HostCommands/BuildIn/FormKeyFromEncryptedComponents_A4.vb
''
'' This program is free software; you can redistribute it and/or modify
'' it under the terms of the GNU General Public License as published by
'' the Free Software Foundation; either version 2 of the License, or
'' (at your option) any later version.
''
'' This program is distributed in the hope that it will be useful,
'' but WITHOUT ANY WARRANTY; without even the implied warranty of
'' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'' GNU General Public License for more details.
''
'' You should have received a copy of the GNU General Public License
'' along with this program; if not, write to the Free Software
'' Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
'' 

Imports ThalesSim.Core.Message
Imports ThalesSim.Core.Cryptography
Imports ThalesSim.Core

Namespace HostCommands.BuildIn

    ''' <summary>
    ''' Forms a key from encrypted components.
    ''' </summary>
    ''' <remarks>
    ''' This class implements the A4 Racal command.
    ''' </remarks>
    <ThalesCommandCode("A4", "A5", "", "Forms a key from encrypted components")> _
    Public Class FormKeyFromEncryptedComponents_A4
        Inherits AHostCommand

        Private _nbrComponents As String = ""
        Private _iNbrComponents As Integer
        Private _keyTypeCode As String = ""
        Private _lmkScheme As String = ""
        Private _comps(8) As String

        ''' <summary>
        ''' Constructor.
        ''' </summary>
        ''' <remarks>
        ''' The constructor sets up the A4 message parsing fields.
        ''' </remarks>
        Public Sub New()
            ReadXMLDefinitions()
        End Sub

        ''' <summary>
        ''' Parses the request message.
        ''' </summary>
        ''' <remarks>
        ''' This method parses the command message. The message header and message command
        ''' code are <b>not</b> part of the message.
        ''' </remarks>
        Public Overrides Sub AcceptMessage(ByVal msg As Message.Message)
            XML.MessageParser.Parse(msg, XMLMessageFields, kvp, XMLParseResult)
            If XMLParseResult = ErrorCodes.ER_00_NO_ERROR Then
                _nbrComponents = kvp.Item("Number of Components")
                _keyTypeCode = kvp.Item("Key Type")
                _lmkScheme = kvp.Item("Key Scheme (LMK)")
                _iNbrComponents = Convert.ToInt32(_nbrComponents)
                For i As Integer = 1 To _iNbrComponents
                    _comps(i - 1) = kvp.Item("Key Component #" + i.ToString)
                Next
            End If
        End Sub

        ''' <summary>
        ''' Creates the response message.
        ''' </summary>
        ''' <remarks>
        ''' This method creates the response message. The message header and message reply code
        ''' are <b>not</b> part of the message.
        ''' </remarks>
        Public Overrides Function ConstructResponse() As Message.MessageResponse
            Dim mr As New MessageResponse

            If Not IsInAuthorizedState() Then
                Log.Logger.MajorInfo("Can't print clear key while not in the AUTHORIZED state")
                mr.AddElement(ErrorCodes.ER_17_HSM_IS_NOT_IN_THE_AUTHORIZED_STATE)
                Return mr
            End If

            Dim LMKKeyPair As LMKPairs.LMKPair, var As Integer
            Dim ks As KeySchemeTable.KeyScheme

            If ValidateKeyTypeCode(_keyTypeCode, LMKKeyPair, var.ToString, mr) = False Then Return mr
            If ValidateKeySchemeCode(_lmkScheme, ks, mr) = False Then Return mr

            Dim clearKeys(8) As String, clearKey As String = ""
            Dim sourceKs As KeySchemeTable.KeyScheme = New HexKey(_comps(0)).Scheme

            For i As Integer = 1 To _iNbrComponents
                clearKeys(i - 1) = Utility.DecryptUnderLMK(_comps(i - 1), sourceKs, LMKKeyPair, var.ToString)
                If Utility.IsParityOK(clearKeys(i - 1), Utility.ParityCheck.OddParity) = False Then
                    mr.AddElement(ErrorCodes.ER_10_SOURCE_KEY_PARITY_ERROR)
                    Return mr
                End If
                If clearKey <> "" Then
                    clearKey = Utility.XORHexStringsFull(clearKey, clearKeys(i - 1))
                Else
                    clearKey = clearKeys(i - 1)
                End If
            Next

            Dim cryptKey As String = Utility.EncryptUnderLMK(clearKey, ks, LMKKeyPair, var.ToString)
            Dim checkValue As String = TripleDES.TripleDESEncrypt(New HexKey(clearKey), ZEROES)

            For i As Integer = 1 To _iNbrComponents
                Log.Logger.MinorInfo("Component " + i.ToString() + " (clear): " + Utility.RemoveKeyType(clearKeys(i - 1)))
            Next
            Log.Logger.MinorInfo("Key (clear): " + clearKey)
            Log.Logger.MinorInfo("Check value: " + checkValue.Substring(0, 6))

            mr.AddElement(ErrorCodes.ER_00_NO_ERROR)
            mr.AddElement(cryptKey)
            mr.AddElement(checkValue.Substring(0, 6))

            Return mr

        End Function

    End Class

End Namespace

File: ThalesSim.Core/Commands/Host/Implementations/FormKeyFromEncryptedComponents_A4.cs
/*
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

using System;
using ThalesSim.Core.Cryptography;
using ThalesSim.Core.Message;
using ThalesSim.Core.Resources;
using ThalesSim.Core.Utility;

namespace ThalesSim.Core.Commands.Host.Implementations
{
    /// <summary>
    /// Thales A4 implementation.
    /// </summary>
    [ThalesHostCommand("A4", "A5", "Forms a key from encrypted components")]
    [AuthorizedState]
    public class FormKeyFromEncryptedComponents_A4 : AHostCommand
    {
        private string _nbrComponents;
        private int _iNbrComponents;
        private string _keyTypeCode;
        private string _lmkScheme;
        private string[] _comps ;

        /// <summary>
        /// Read XML definitions on instantiation.
        /// </summary>
        public FormKeyFromEncryptedComponents_A4()
        {
            ReadXmlDefinitions();
        }

        /// <summary>
        /// Accept message from client.
        /// </summary>
        /// <param name="message">Request message.</param>
        public override void AcceptMessage(StreamMessage message)
        {
            base.AcceptMessage(message);

            if (XmlParseResult != ErrorCodes.ER_00_NO_ERROR)
            {
                return;
            }

            _nbrComponents = KeyValues.Item("Number of Components");
            _keyTypeCode = KeyValues.Item("Key Type");
            _lmkScheme = KeyValues.Item("Key Scheme (LMK)");
            _iNbrComponents = Convert.ToInt32(_nbrComponents);
            _comps = new string[_iNbrComponents];
            for (var i = 1; i <= _iNbrComponents; i++)
            {
                _comps[i - 1] = KeyValues.ItemCombination("Key Component Scheme #" + i.ToString(),
                                                          "Key Component #" + i.ToString());
            }
        }

        /// <summary>
        /// Process request.
        /// </summary>
        /// <returns>Response message.</returns>
        public override StreamResponse ConstructResponse()
        {
            var mr = new StreamResponse();

            KeyTypeCode ktc = null;
            if (!ValidateKeyTypeCode(_keyTypeCode, mr, ref ktc))
            {
                return mr;
            }

            var lmkKs = KeyScheme.Unspecified;
            if (!ValidateKeySchemeCode(_lmkScheme, mr, ref lmkKs))
            {
                return mr;
            }

            var clearComps = new HexKeyThales[_iNbrComponents];
            var clearKey = string.Empty;
            for (var i = 1; i <= _iNbrComponents; i++)
            {
                clearComps[i - 1] = new HexKeyThales(ktc, false, _comps[i - 1]);
                if (!clearComps[i - 1].ClearKey.IsParityOk(Parity.Odd))
                {
                    mr.Append(ErrorCodes.ER_10_SOURCE_KEY_PARITY_ERROR);
                    return mr;
                }

                clearKey = !string.IsNullOrEmpty(clearKey) ? clearKey.XorHex(clearComps[i - 1].ClearKey) : clearComps[i - 1].ClearKey;
            }

            clearKey = clearKey.MakeParity(Parity.Odd);
            var cryptKey = new HexKeyThales(ktc, true, clearKey);

            for (var i = 1; i <= _iNbrComponents; i++)
            {
                Log.InfoFormat("Component {0} (clear): {1}", i, clearComps[i - 1].ClearKey);
            }
            Log.InfoFormat("Key (clear): {0}", clearKey);
            Log.InfoFormat("Check value: {0}", cryptKey.CheckValue);

            mr.Append(ErrorCodes.ER_00_NO_ERROR);
            if (lmkKs == KeyScheme.DoubleLengthKeyVariant || lmkKs == KeyScheme.TripleLengthKeyVariant)
            {
                mr.Append(cryptKey.KeyVariant);
            }
            else
            {
                mr.Append(cryptKey.KeyAnsi);
            }
            mr.Append(cryptKey.CheckValue.Substring(0, 6));

            return mr;
        }
    }
}
